0.2 Structure shifts

AllaYakymova opened the pull request:
1. Reorganized folder structure. 
2. Moved Joi validation to separate folder, 
3. Wrapped controller methods in asynchandler 
4. Created errors 
5. Reorganized validation middlewares 
6. 6 changed basic routes in app.ts 
7. Created admin route

AllaYakymova committed:
0.2 Structure shifts

kolesnicknick commented on Dp-202-shop/server/src/app.ts:
```diff
@@ -16,10 +16,22 @@ sequelize.authenticate()
+
+
+app.use('/products', productsRouter);
+// app.use('/order',  authMiddleware, ordersRouter);
```
remove comment

kolesnicknick commented on Dp-202-shop/server/src/app.ts:
```diff
@@ -16,10 +16,22 @@ sequelize.authenticate()
+// app.use('/order',  authMiddleware, ordersRouter);
+app.use('/admin', /*authMiddleware, adminMiddleWare */ adminRouter);
+
+app.use(function (req, res, next) {
+  next(new NotFound('This page doesn\'t exist'))
+});
```
put this middleware in a separate file

kolesnicknick commented on Dp-202-shop/server/src/common/dtos/new.product.dto.ts:
```diff
@@ -0,0 +1,46 @@
+}
+
+
+export const newProductDto = Joi.object({
+    product: Joi.object({ 
+    vendor_code: Joi.string().required(),
+    product_name: Joi.string().required(),
+    manufacture: Joi.string().required(),
+    category: Joi.string().required().allow(''),
+    unit: Joi.string().required().allow(''),
+    amount: Joi.string().min(1).required(),
+    price: Joi.number().required(),
+    deleted: Joi.boolean().required(),
+    ingredients: Joi.string().required().allow(''),
+    img: Joi.string().required().allow(''),
+    createdAt: Joi.date(),
+    updatedAt: Joi.date(),
+    user_id: Joi.number().required()
+    })
+});
```
if you allow('') it means your field is not required, change it

valeriia-danylenko replied:
> if you allow('') it means your field is not required, change it

But what if a person wants to delete this field?

kolesnicknick commented on Dp-202-shop/server/src/common/dtos/search.params.dto.ts:
```diff
@@ -0,0 +1,17 @@
+    products?: string
+}
+
+export const searchParamsDto = Joi.object().keys({
+    categories: Joi.string().pattern(/^[0-9,]+$/).allow(''),
+    manufactures: Joi.string().allow(''),
+    products: Joi.string().allow('')
+}).unknown(true);
```
why do we allow('') everywhere?

valeriia-danylenko replied:
> why do we allow('') everywhere?

Because I wanted /search to work even without parameters

kolesnicknick commented on Dp-202-shop/server/src/common/dtos/new.product.dto.ts:
```diff
@@ -0,0 +1,46 @@
+    };
+}
+
+interface Product {
```
IProduct, if you name interfaces in this way everywhere

kolesnicknick commented on Dp-202-shop/server/src/common/errors/api-errors.ts:
```diff
@@ -1,18 +1,20 @@
     super();
+    this.data = data;
   }
 
-  static badRequest(message: string): ApiError {
-    return new ApiError(404,  message);
+  static badRequest(message: string, data: any[],): ApiError {
+    return new ApiError(404, message, data);
   }
 
-  static internal(message): ApiError {
-    return new ApiError(500, message)
+  static internal(message: string, data: any[]): ApiError {
+    return new ApiError(500, message, data)
   }
 
-  static unauthorized(message: string): ApiError {
-    return new ApiError(401, message);
+  static unauthorized(data: any[], message: string): ApiError {
+    return new ApiError(401, message, data);
   }
+
 }
```
change ALL errors to different classes, not just static methods

iseldrewGD replied:
Please, stick to one naming convention. 
`apiErrors.ts`

kolesnicknick commented on Dp-202-shop/server/src/common/helpers/dataNormalization.js:
```diff
@@ -0,0 +1,69 @@
+    if (rawObj.dataValues) {
+        obj = rawObj.dataValues;
+    }
+    if (columns.length === 0) return obj;
+    columns = columns[0];
+
+    const valuesToExtract = [];
+    columns.forEach(column => recursion(valuesToExtract, column));
+    return extractDataObj(valuesToExtract, obj);
+};
+
+
+const extractDataValues = function (rawData) {
+    return rawData.map(el => el.dataValues);
+};
+
+const recursion = function (result, object, ...prevRound) {
+    for (const [key, value] of Object.entries(object)) {
+        prevRound.push(key);
+        value.forEach(val => {
+            if (typeof val === 'object' && val !== null) {
+                return recursion(result, val, ...prevRound)
+            }
+            result.push({ value: val, path: [...prevRound, val].join('.dataValues.').split('.') })
+        })
+    }
+};
+
+export const delExtra = function (rawData, keys) {
+   return rawData.map(el => deleteKeys(el, keys));
+};
+
+export const deleteKeys = function (el, keys) {
+    keys.forEach(key => {
+        delete el[key]
+    });
+    return el
+};
+
+// module.exports = { delExtra, normalize, normalizeOne, deleteKeys };
```
great code. Can we use somehow library lodash to simplify this file?

kolesnicknick commented on Dp-202-shop/server/src/common/middleware/auth.middleware.ts:
```diff
@@ -1,18 +1,23 @@
-    // next();
-  // } catch (e) {
-  //   next(ApiError.internal(e.message));
-  // }
+  const { phone, password } = req.headers;
+  const foundUser = await userService.findUserByPhone(phone);
+  if (foundUser.length === 0) {
+    res.locals.isAuthenticated = false;
+    next(new NotFound('Such phone number is not registered'));
+  } else if (foundUser[0].password != password) {
+    res.locals.isAuthenticated = false;
+    next(new Forbidden('Incorrect password'));
+  } else {
+    res.locals.user = foundUser[0];
+    res.locals.isAuthenticated = true;
+    next();
+  }
 }
```
is this one should be done by Anastasia's PR?

kolesnicknick commented on Dp-202-shop/server/src/common/middleware/check.products.middleware.ts:
```diff
@@ -0,0 +1,22 @@
+import { IProductRequestSchema } from '../dtos/orders.dto';
+
+
+export const checkProductsMiddleware = async function (req: ValidatedRequest<IProductRequestSchema>, res: Response, next: NextFunction): Promise<void> {
```
better use arrow functions

kolesnicknick commented on Dp-202-shop/server/src/modules/admin/admin.controller.ts:
```diff
@@ -0,0 +1,41 @@
+
+    addNewProduct = asyncHandler(async (req: ValidatedRequest<INewProduct>, res: Response, next: NextFunction): Promise<void> => {
+        const { product } = req.body;
+        console.log(product);
```
remove console log

iseldrewGD commented on Dp-202-shop/server/src/modules/users/user.service.ts:
```diff
@@ -0,0 +1,77 @@
+  };
+
+  async findUserByPhone(phoneNum) {
+    const rawUser = await User.findAll({
```
In case where you need to find one exact user use `.findOne()`

iseldrewGD commented on Dp-202-shop/server/src/modules/admin/admin.router.ts:
```diff
@@ -0,0 +1,15 @@
+
+adminRouter.get("/products/add", /*hashCheckMiddlware, */ adminController.showNothing);
+adminRouter.post("/products/add", /*hashCheckMiddlware, */ validator.body(newProductDto), adminController.addNewProduct);
+adminRouter.get("/allsalesmen",  adminController.getAllSalesmen);
```
REST routes should be in kebab case `/all-salesmen`

iseldrewGD commented on Dp-202-shop/server/src/modules/products/product.router.ts:
```diff
@@ -1,9 +1,15 @@
+productsRouter.get("/search", validator.query(searchParamsDto), productsController.getAll);
+productsRouter.get("/:id", validator.params(idDto), productsController.getOne);
+productsRouter.get("/", productsController.getByQuery);
+productsRouter.get('/allcategories', productsController.getAllCategories);
+productsRouter.get('/allmanufactures', productsController.getAllManufactures);
+productsRouter.get('/allunits', productsController.getAllUnits);
```
REST routes should be in kebab case

iseldrewGD commented on Dp-202-shop/server/src/modules/users/user.service.ts:
```diff
@@ -0,0 +1,77 @@
+    return normalizeOne(newUser).id
+  }
+  async findOrCreateSalesman(login) {
+    const rawSalesman = await User.findAll({
```
Use `.findOne()`

iseldrewGD commented on Dp-202-shop/server/src/modules/users/user.service.ts:
```diff
@@ -0,0 +1,77 @@
+    const rawSalesmenData = await User.findAll({
+      attributes:
+        ['id', 'login'],
+      where: { role: role }
```
ES6 syntax allows us to simplify it: `where: { role }`

valeriia-danylenko committed:
Applied all remarks

Deleted redundant ApiError.ts
modified newProductDto
modified searchParamsDto
removed some redundant commented code
created notFoundMiddleware for all routes that do not exist
Renamed Rest routes using kebab-case

valeriia-danylenko committed:
Update view.ts

AllaYakymova committed:
0.2 Structure shifts some fixes in models interfaces of OrderItem & Order

AllaYakymova committed:
Merge branch '0.2_structure_shift' of https://github.com/ITA-Dnipro/Dp-202-shop into 0.2_structure_shift
